torchrl/envs/model_based.py

```diff
@@ -0,0 +1,94 @@
+from .common import EnvBase
+
+
+class ModelBasedEnv(EnvBase, metaclass=abc.ABCMeta):
+    """
+    Basic environnement for Model Based RL algorithms.
+
+    This class is a wrapper around the model of the MBRL algorithm.
+    This class is meant to give an env framework to a world model and a reward model.
+    It is meant to behave as a classical environment.
+
+    Properties:
+        - observation_spec (CompositeSpec): sampling spec of the observations;
+        - action_spec (TensorSpec): sampling spec of the actions;
+        - input_spec (CompositeSpec): sampling spec of the actions and/or other inputs;
+        - reward_spec (TensorSpec): sampling spec of the rewards;
+        - batch_size (torch.Size): number of environments contained in the instance;
+        - device (torch.device): device where the env input and output are expected to live
+        - is_done (torch.Tensor): boolean value(s) indicating if the environment has reached a done state since the
+            last reset
+
+    Args:
+        world_model (nn.Module): model that generates world states and its corresponding rewards;
+        device (torch.device): device where the env input and output are expected to live
+        dtype (torch.dtype): dtype of the env input and output
+        batch_size (torch.Size): number of environments contained in the instance
+
+    Methods:
+        step (TensorDict -> TensorDict): step in the environment
+        reset (TensorDict, optional -> TensorDict): reset the environment
+        set_seed (int -> int): sets the seed of the environment
+        rand_step (TensorDict, optional -> TensorDict): random step given the action spec
+        rollout (Callable, ... -> TensorDict): executes a rollout in the environment with the given policy (or random
+            steps if no policy is provided)
+
+    """
+
+    def __init__(
+        self,
+        world_model: Union[nn.Module, List[TensorDictModule], TensorDictSequence],
+        device: DEVICE_TYPING = "cpu",
+        dtype: Optional[Union[torch.dtype, np.dtype]] = None,
+        batch_size: Optional[torch.Size] = None,
+    ):
+        super(ModelBasedEnv, self).__init__(
+            device=device, dtype=dtype, batch_size=batch_size
+        )
+        self.world_model = world_model
+
+    def set_specs_from_env(self, env: EnvBase):
+        """
+        Sets the specs of the environment from the specs of the given environment.
+        """
+        self.observation_spec = env.observation_spec
+        self.action_spec = env.action_spec
+        self.reward_spec = env.reward_spec
+
+    def _step(
+        self,
+        tensordict: TensorDict,
+    ) -> TensorDict:
+        # step method requires to be immutable
+        tensordict_out = tensordict.clone()
+        # Compute world state
+        tensordict_out = self.world_model(tensordict_out)
+        # Step requires a done flag. No sense for MBRL so we set it to False
```

Comment:
I'm not sure this statement is generally true. There are some MBRL methods where the model also predicts termination conditions which are used by the policies.

Reply:
Ok so you would say that it's better to let the world_model to handle it? Or having a method that  from tensordict out creates the done tensor? Do you have a paper in mind so i can read it to understand the logic behind this? Thank you for the help!

Reply:
Sometimes the world model itself would handle a learned termination condition. [Here](http://proceedings.mlr.press/v144/amos21a/amos21a.pdf) is a recent example (see Eq. 12). For example, one could establish some convention that if the tensordict coming from the model has a "done" tag you would use it, otherwise, you set it to a default value.

BTW, in other methods sometimes not everything is learned. For example, in PETS the authors assume that the reward function is known, and in MBPO the termination function is also known. In MBRL-Lib we [allow](https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/models/model_env.py#L42) users to optionally pass arbitrary tensors -> tensor functions that can be used to calculate rewards/termination for a given observation and action values. 

Here this could maybe be taken care of by each `ModelBasedEnv` subclass, with some abstractions defined by the base class?

Reply:
I changed the "done" part such as world model can predict it. Thanks for the help!

Reply:
For the reward part i don 't think the current MBEnv is incompatible with what you are describing, indeed, this can be handled by the definition of the world model. You just need to define your reward function as a TDModule that have require_grad = False and then wrap it with the rest of the model part. If you think it's better that MBEnv takes world modeler and reward as an input instead i can make the change. Tell me what you think!

Reply:
I guess it depends a bit on the philosophy of the library. For MBRL-Lib we wanted to make things very modular, and in fact rely a lot on Hydra to let users swap things around easily. That's why it made sense to me to provide these as separate arguments and let the user just change which option to use via config files, rather than having them modify their model code. 

It seems to me that TorchRL is more lower level oriented (as in, here are well-tested and efficient structures and utilities that should make life easier for users), so I think current design is OK as long as this functionality can be achieved, which I agree should be the case. But I'm also curious to hear what @vmoens thinks.

Reply:
A generic MBRL env class will have to predict a few things: observations, reward, done state, safety measure, possibly other things too.
I understand your point @luisenp that asking users to change their model can reduce modularity in some instances, but on the high level i think having just one model that is responsible of executing the simulation step makes it more easy to generalize the class over algorithms.
Say we build the class like this
```python
class ModelBasedEnv(EnvBase, metaclass=abc.ABCMeta):
    def __init__(
        self,
        obs_mode,
        reward_model,
        done_model,
        safety_model):
        foo()
```
I'm concerned that tomorrow someone comes and says: hey guys, how can I fit a new model that predicts something else. Then we'll either have to tell this person how to code her own env from scratch or refactor our implementation for a single use case.
I think that here we can get out of this by carefully thinking about the general use case and precisely documenting how to build a model-based environment that has or has not a model for reward, done, safety etc.
I think we can do all of this with just a few lines of code and the `TensorDictModule` will allow you to recycle most of the code across experiments.
The way I see it, it should look like this

```
# modules
obs_model = TensorDictModule(obs_net, in_keys=["observation", "action"], out_keys=["next_observation"])
reward_model = TensorDictModule(reward_net, in_keys=["observation", "action"], out_keys=["reward"])
done_model = TensorDictModule(done_net, in_keys=["observation", "action"], out_keys=["done"])
safety_model = TensorDictModule(safety_net, in_keys=["observation", "action"], out_keys=["safety_measure"])

# we can build an env with or without done
env1 = ModelBasedEnv(TensorDictSequence(obs_model, reward_model, safety_model))  # will assume done is False
env2 = ModelBasedEnv(TensorDictSequence(obs_model, reward_model, done_model, safety_model))  # will compute done from obs

# we can build an env with or without reward
env3 = ModelBasedEnv(TensorDictSequence(obs_model, safety_model))  # in this case the output will miss the reward key, has it does not have a default value

```
(by the way, safety net is a great name haha)

Does that make sense?
I feel this would work with a hydra config easily but IMO it's still highly modular (perhaps I misunderstand what you mean by modular, in which case I'm interested to know what it would take for it to satisfy that).

On the "con" side it's true that this class is not super explicit but, again, nothing can't be done with a good doc.

@nicolas-dufour 
we could move model_based.py to a model_based directory, and put a `readme.md` where we give a simple example of how that works, what do you think?

Reply:
Yep this sound good!

Reply:
Hi @vmoens! I wonder if we were talking about different things? I pretty much agree with your vision of having a general model design and let the user decide what to provide or not for their use case via `TensorDictModule`. This completely makes sense to me. 

In my last comment I was referring specifically to letting the user have an easy way to pass custom reward (or termination) functions that are not necessarily considered part of the model (e.g. pass the ground truth reward, or no reward, or whatever), and swap ones for others. But this is not the same as suggesting that separate inputs should be passed for every possible submodel, which is maybe what my comment was interpreted as? The model can still provide arbitrary outputs, but, for example, if a hard-coded reward is given, then any (optional) reward predictions returned by the model are just ignored.

Moreover, since in `mbrl` we already provide a bunch of utilities building from pre-defined Hydra configs, a user can set up something like "use PlaNet state-based model to solve gym-HalfCheetah but actually run it with the ground truth reward function available here", and they can do this via command line with no changes to the model code. If they want to try a new reward function, they can just add another tensor->tensor function to a module that can be imported, and point to it with a simple command line argument. This is what I meant by modularity. I do understand that this can already be achieved in current `torch-RL` by putting these custom hard-coded functions inside the world model, but it seems to me that this would require some extra boilerplate?

But, in any case, I'm not arguing that what we did is the right way to do things, just that it made sense for what we wanted to do with `mbrl` (which was to have plug-and-play components with Hydra as a glue for everything).

Reply:
Thanks for the clarification!
Indeed I think I lack a bit of context there :) . Would you have something to point me to for me to get a better understanding of what it would take for us to make this possible?

Reply:
[This](https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/models/model_env.py#L42) is what we did in `mbrl`, and here is an example [config](https://github.com/facebookresearch/mbrl-lib/blob/main/mbrl/examples/conf/overrides/pets_halfcheetah.yaml#L3) that specifies the termination function to use. The rewards_fn is not specified because we also provide some defaults in `utils` package to use when `learned_rewards: false`. Also note that, in principle, `termination_fn` should also be an optional kwarg, but when I first wrote `ModelEnv` I hand't consider that termination functions could be learned. 

I guess the question for `torchRL` is how do you want to support non-learned components of the model. If the philosophy of the library is to provide a flexible API but let users figure out how to configure things, then it seems fine to me that they just code these into the world model---and I think this is already how it works, if I understand correctly.